utils/abrt-action-analyze-java.1
'\" t
.\"     Title: abrt-action-analyze-java
.\"    Author: [see the "AUTHORS" section]
.\"      Date: 01/19/2014
.\"    Manual: ABRT Manual
.\"  Language: English
.\"
.TH "ABRT\-ACTION\-ANALYZ" "1" "01/19/2014" "abrt-java-connector" "ABRT Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
abrt-action-analyze-java \- Calculate and save UUID & DUPHASH and determine the level of usability for reporting of a Java stack trace\&.
.SH "SYNOPSIS"
.sp
\fIabrt\-action\-analyze\-java\fR [\-v] [\-d DIR] [\-o] [\-f FILE]
.SH "DESCRIPTION"
.sp
The tool reads the file named \fIbacktrace\fR from a problem data directory, processes it and generates a universally unique identifier (UUID)\&. Then it saves this data as new element \fIuuid\fR\&. It also checks whether the stack trace contains a remote address in any of its frames and if so it creates \fInot-reportable\fR element whose contents explains why the stack trace should not be reported into a bug tracking system\&.
.SS "Integration with ABRT events"
.sp
\fIabrt\-action\-analyze\-java\fR can be used to generate the UUID & DUPHAS of a newly saved Java stack trace\&.
.sp
.if n \{\
.RS 4
.\}
.nf
EVENT=post\-create analyzer=Java   abrt\-action\-analyze\-java
.fi
.if n \{\
.RE
.\}
.SH "OPTIONS"
.PP
\-d DIR
.RS 4
Path to a problem directory\&. The tool reads the backtrace from stdin when neither this option nor -f is provided\&.
.RE
.PP
\-f FILE
.RS 4
Path to a stack trace\&. The tool reads the backtrace from stdin when neither this option nor -d is provided\&.
.RE
.PP
\-o
.RS 4
Print the result to stdout\&.
.RE
.PP
\-v
.RS 4
Be more verbose\&. Can be given multiple times\&.
.RE
.SH "AUTHORS"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
ABRT team
.RE


ChangeLog
2013-05-30  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/RemoteTest.java:
	Added missing test class.

2013-05-22  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Solution for JAR-style URIs:
	"file:/home/tester/abrt_connector/bin/JarTest.jar!/SimpleTest.class"

2013-05-21  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added code which acquire name and path to a main class.

2013-05-20  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Basic support for acquiring main class name and using that name
	as "executable" attribute in event send ABRT.

2013-04-30  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Changed part of code for reading executable name. Name is now read
	from link /proc/${PID}/exe.

2013-04-29  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Changed names of attributes which are send to ABRT. Those names can't
	contain dots ('.') - replaced by underscores ('_'). Renamed a few of
	other attributes.

2013-04-26  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Implemented new function used to format class signature into a form
	suitable for ClassLoader.getResource(). Fixed stack trace list by
	using this function.

2013-04-25  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* Makefile:
	Fixed 'clean' rule to really clean all generated files (classes and Jars).

2013-04-25  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/com/redhat/abrt/test/Test.java:
	Added new test class and its package.
	* Makefile:
	Added rule to compile and run new test class.
	Updated 'clean' rule.
	* TODO:
	Fixed typo.

2013-04-22  Jakub Filak  <jfilak@redhat.com>

	* src/abrt-checker.c:
	ABRT's own field names are used instead of constants defined in
	abrt-checker.c.

2013-04-22  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Stack trace is now saved into the log file too.

2013-04-02  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Paths to class files and java archives are now printed in
	the generated stack trace.

2013-03-29  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Proper cleanup code added into many functions. Refactored stack
	trace into its own function.

2013-03-28  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Minor update: unification of names of common local variables and
	function parameters.

2013-03-27  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Agent is now able to acquire and report basic process properties -
	command used to start JVM and command line parameters (incl. -Xfoo=bar
	ones).

2013-03-26  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Fixed behaviour of REPORT_ERRORS_TO_ABRT macro, added two macros
	used together with __LINE__ macro, change jvmEnvironment struct
	into new data type, code refactoring.

2013-01-30  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added basic support for printing path to resource (.class, .jar).

2013-01-28  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added support for retrieving line numbers for exception stack traces.

2013-01-25  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Finished code used for sending JVM environment variables to ABRT tool.

2013-01-21  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Comments enhancements, basic support for sending JVM environment
	variables to ABRT tool.

2013-01-18  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added three JVM environment variables which needs to be read and sent
	to ABRT tool.

2012-12-03  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Refactored stack trace generation code into its own print_stack_trace()
	function. Changed formating of stack trace - now it is similar to
	Exception.printStackTrace(). Fixed supposed memory leak in this
	part of code.

2012-12-03  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Updated functions used to print class name and method signature.

2012-11-28  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added function which can format class signature
	as a printable class name.

2012-11-15  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added support for more JVM environment variables.
	#define to enable/disable ABRT reporting.

2012-11-13  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Refactoring - get rid of two global variables.

2012-11-12  Jakub Filak  <jfilak@redhat.com>

	* Makefile:
	Fixed problem with output directory.

2012-11-01  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added support for checking up JVM environment and
	storing these info to a structure.

2012-11-01  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added function to get UID, minor refactoring.

2012-10-29  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* TODO:
	Added three items to the TODO list.

2012-10-29  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* src/abrt-checker.c:
	Added source code of ABRT checker.
	* src/SimpleTest.java:
	* src/Test.java:
	Added source code of two tests.
	* Makefile:
	Created project makefile
	* LICENSE:
	Changed project name.

2012-10-29  Pavel Tisnovsky  <ptisnovs@redhat.com>

	* ChangeLog:
	* AUTHORS:
	* LICENCE:
	* NEWS:
	* README:
	* TODO:
	Created.



utils/abrt-action-analyze-java.c
/*
    Copyright (C) 2014  Red Hat, Inc.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include <satyr/location.h>
#include <satyr/thread.h>
#include <satyr/java/stacktrace.h>
#include <satyr/java/thread.h>
#include <satyr/java/frame.h>

#include <abrt/libabrt.h>
#include <stdlib.h>

/* 4 = 1 exception + 3 methods */
#define FRAMES_FOR_DUPHASH 4

typedef struct
{
    const char *name;
    char *data;
    int nofree;
} analysis_result_t;

static char *
backtrace_from_dump_dir(const char *dir_name)
{
    struct dump_dir *dd = dd_opendir(dir_name, DD_OPEN_READONLY);
    if (NULL == dd)
    {
        return NULL;
    }

    /* Read backtrace */
    /* Prints an error message if the file cannot be loaded */
    char *backtrace_str = dd_load_text_ext(dd, FILENAME_BACKTRACE,
                                           DD_LOAD_TEXT_RETURN_NULL_ON_FAILURE);

    dd_close(dd);

    return backtrace_str;
}

static void
write_results_to_dump_dir(const char *dir_name,
        const analysis_result_t *res_begin, const analysis_result_t *res_end)
{
    struct dump_dir *dd = dd_opendir(dir_name, /*Open for writing*/0);
    if (NULL != dd)
    {
        const analysis_result_t *res = res_begin;

        for ( ; res != res_end; ++res)
            dd_save_text(dd, res->name, res->data);

        dd_close(dd);
    }
}

static void
write_to_fd(int fdout, const char *message)
{
    full_write(fdout, message, strlen(message));
    full_write(fdout, "\n", 1);
}


static void
write_results_to_fd(int fdout,
        const analysis_result_t *res_begin, const analysis_result_t *res_end)
{
    const analysis_result_t *res = res_begin;

    for ( ; res != res_end; ++res)
    {
        write_to_fd(fdout, res->name);
        write_to_fd(fdout, res->data);
    }
}

static void
write_results_to_file(const analysis_result_t *res_begin, const analysis_result_t *res_end)
{
    const analysis_result_t *res = res_begin;

    for ( ; res != res_end; ++res)
    {
        int fdout = open(res->name,
                O_WRONLY | O_TRUNC | O_CREAT | O_NOFOLLOW,
                S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP );

        if (0 > fdout)
        {
            perror_msg("Can't open file '%s' for writing", res->name);
            continue;
        }

        write_to_fd(fdout, res->data);
        close(fdout);
    }
}

static char *
backtrace_from_fd(int fdin)
{
    return xmalloc_read(fdin, /*no size limit*/NULL);
}

static char *
backtrace_from_file(const char *file_name)
{
    return xmalloc_xopen_read_close(file_name, /*no size limit*/NULL);
}

static char *
work_out_list_of_remote_urls(struct sr_java_stacktrace *stacktrace)
{
    struct strbuf *remote_files_csv = strbuf_new();
    struct sr_java_thread *thread = stacktrace->threads;
    while (NULL != thread)
    {
        struct sr_java_frame *frame = thread->frames;
        while (NULL != frame)
        {
            if (NULL != frame->class_path && prefixcmp(frame->class_path, "file://") != 0)
            {
                struct stat buf;
                if (stat(frame->class_path, &buf) && errno == ENOENT)
                {
                    if (strstr(remote_files_csv->buf, frame->class_path) == NULL)
                    {
                        log_debug("Adding a new path to the list of remote paths: '%s'", frame->class_path);
                        strbuf_append_strf(remote_files_csv, "%s%s",
                                remote_files_csv->buf[0] != '\0' ? ", " : "",
                                frame->class_path);
                    }
                    else
                        log_debug("The list of remote paths already contains path: '%s'", frame->class_path);
                }
                else
                    log_debug("Class path exists or is malformed: '%s'", frame->class_path);
            }
            frame = frame->next;
        }
        thread = thread->next;
    }

    if (remote_files_csv->buf[0] != '\0')
    {
        return strbuf_free_nobuf(remote_files_csv);
    }

    strbuf_free(remote_files_csv);
    return NULL;
}

int main(int argc, char *argv[])
{
#if ENABLE_NLS
    /* I18n */
    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);
#endif

    abrt_init(argv);

    const char *dump_dir_name = NULL;
    const char *backtrace_file = NULL;

    /* Can't keep these strings/structs static: _() doesn't support that */
    const char *program_usage_string = _(
        "& [[-d DIR] | [-f FILE]] [-o]\n"
        "\n"
        "Analyzes Java backtrace, generates duplication hash and creates\n"
        "not-reportable file for bracktraces whose frames have remote files in their\n"
        "class path\n"
    );
    enum {
        OPT_v = 1 << 0,
        OPT_d = 1 << 1,
        OPT_f = 1 << 2,
        OPT_o = 1 << 3,
    };
    /* Keep enum above and order of options below in sync! */
    struct options program_options[] = {
        OPT__VERBOSE(&g_verbose),
        OPT_STRING('d', "dumpdir", &dump_dir_name, "DIR", _("Problem directory")),
        OPT_STRING('f', "backtrace", &backtrace_file, "FILE", _("Path to backtrace")),
        OPT_BOOL('o', "stdout", NULL, _("Print results on standard output")),
        { 0 }
    };
    program_options[ARRAY_SIZE(program_options) - 1].type = OPTION_END;

    unsigned opts = parse_opts(argc, argv, program_options, program_usage_string);

    export_abrt_envvars(0);

    if (NULL != dump_dir_name && NULL != backtrace_file)
        error_msg_and_die("You need to pass either DIR or FILE");

    int retval = 1;
    char *backtrace_str = NULL;
    if (NULL != dump_dir_name)
    {
        backtrace_str = backtrace_from_dump_dir(dump_dir_name);
    }
    else if (NULL != backtrace_file)
    {
        backtrace_str = backtrace_from_file(backtrace_file);
    }
    else
    {
        backtrace_str = backtrace_from_fd(STDIN_FILENO);
    }

    if (NULL == backtrace_str)
        goto finish;

    struct sr_location location;
    sr_location_init(&location);
    const char *backtrace_str_ptr = backtrace_str;
    struct sr_java_stacktrace *stacktrace = sr_java_stacktrace_parse(&backtrace_str_ptr, &location);
    free(backtrace_str);

    if (NULL == stacktrace)
    {
        error_msg("Could not parse the stack trace");
        goto finish;
    }

    analysis_result_t results[3] = { { 0 } };
    analysis_result_t *results_iter = results;

    char *remote_files_csv = work_out_list_of_remote_urls(stacktrace);

    char *hash_str = NULL;
    struct sr_thread *crash_thread = (struct sr_thread *)stacktrace->threads;
    if (g_verbose >= 3)
    {
        hash_str = sr_thread_get_duphash(crash_thread, FRAMES_FOR_DUPHASH,
                /*noprefix*/NULL, SR_DUPHASH_NOHASH);
        log("Generating duphash from string: '%s'", hash_str);
        free(hash_str);
    }

    hash_str = sr_thread_get_duphash(crash_thread, FRAMES_FOR_DUPHASH,
            /*noprefix*/NULL, SR_DUPHASH_NORMAL);

    /* DUPHASH is used for searching for duplicates in Bugzilla */
    results_iter->name = FILENAME_DUPHASH;
    results_iter->data = hash_str;
    ++results_iter;

    /* UUID is used for local deduplication */
    results_iter->name = FILENAME_UUID;
    results_iter->data = hash_str;
    results_iter->nofree = 1;
    ++results_iter;

    sr_java_stacktrace_free(stacktrace);

    if (NULL != remote_files_csv)
    {
        results_iter->name = FILENAME_NOT_REPORTABLE;
        results_iter->data = xasprintf(
        _("This problem can be caused by a 3rd party code from the "\
        "jar/class at %s. In order to provide valuable problem " \
        "reports, ABRT will not allow you to submit this problem. If you " \
        "still want to participate in solving this problem, please contact " \
        "the developers directly."), remote_files_csv);
        ++results_iter;
        free(remote_files_csv);
    }

    if (opts & OPT_o)
    {
        write_results_to_fd(STDOUT_FILENO, results, results_iter);
    }
    else if (NULL != dump_dir_name)
    {
        write_results_to_dump_dir(dump_dir_name, results, results_iter);
    }
    else
    {   /* Just write it to the current working directory */
        write_results_to_file(results, results_iter);
    }

    const analysis_result_t *res = results;
    for (; res != results_iter; ++res)
    {
        if (!res->nofree)
        {
            free(res->data);
        }
    }

    retval = 0;
finish:

    return retval;
}
